File: bindings/python/examples/tool_calling.py
#!/usr/bin/env python3
"""
Tool calling example using luup-agent Python bindings.

This example demonstrates:
- Registering custom tools with the @agent.tool() decorator
- Automatic tool invocation by the agent
- Type hints for parameter schema generation
"""

import sys
from pathlib import Path
from typing import Optional

from luup_agent import Model, Agent


def main():
    print("luup-agent Python Tool Calling Example")
    print("=" * 50)
    
    # Check for model path
    if len(sys.argv) < 2:
        print("Usage: python tool_calling.py <path-to-model.gguf>")
        print("Example: python tool_calling.py ../../models/qwen2-0.5b-instruct-q4_k_m.gguf")
        return 1
    
    model_path = sys.argv[1]
    
    if not Path(model_path).exists():
        print(f"Error: Model file not found: {model_path}")
        return 1
    
    print(f"\nCreating model from: {model_path}")
    
    # Create model
    model = Model.from_local(
        model_path,
        gpu_layers=-1,
        context_size=2048,
        threads=0,
    )
    
    print("Model created successfully!")
    
    # Create agent with tool calling enabled
    agent = Agent(
        model,
        system_prompt=(
            "You are a helpful assistant with access to tools. "
            "Use the get_weather tool to check weather and "
            "the calculate tool for math operations. "
            "Always use tools when appropriate."
        ),
        temperature=0.7,
        max_tokens=512,
        enable_tool_calling=True,
        enable_history=True
    )
    
    print("Agent created successfully!\n")
    
    # Register weather tool
    @agent.tool(description="Get current weather for a city")
    def get_weather(city: str, units: str = "celsius") -> dict:
        """Get current weather information for a city."""
        print(f"  [Tool called: get_weather(city='{city}', units='{units}')]")
        
        # In a real implementation, you would call a weather API
        # For this example, return mock data
        return {
            "city": city,
            "temperature": 72,
            "condition": "sunny",
            "humidity": 45,
            "units": units
        }
    
    # Register calculator tool
    @agent.tool(description="Perform mathematical calculations")
    def calculate(expression: str) -> dict:
        """Evaluate a mathematical expression."""
        print(f"  [Tool called: calculate(expression='{expression}')]")
        
        try:
            # WARNING: eval() is dangerous in production!
            # Use a proper math parser in real applications
            result = eval(expression, {"__builtins__": {}}, {})
            return {"result": result, "expression": expression}
        except Exception as e:
            return {"error": str(e), "expression": expression}
    
    # Register user info tool
    @agent.tool(name="get_user_info", description="Get information about a user")
    def get_user(user_id: int) -> dict:
        """Get user information by ID."""
        print(f"  [Tool called: get_user_info(user_id={user_id})]")
        
        # Mock user database
        users = {
            1: {"name": "Alice", "role": "Engineer", "active": True},
            2: {"name": "Bob", "role": "Designer", "active": True},
            3: {"name": "Charlie", "role": "Manager", "active": False},
        }
        
        return users.get(user_id, {"error": "User not found"})
    
    print("Tools registered:")
    print(f"  - {len(agent._tools)} tools available")
    for tool_name in agent._tools.keys():
        print(f"    • {tool_name}")
    print()
    
    print("=" * 50)
    print("Example queries:")
    print("  - What's the weather in Seattle?")
    print("  - Calculate 42 * 13")
    print("  - What's the weather in Tokyo and London?")
    print("  - Tell me about user 1")
    print("=" * 50)
    print()
    
    # Test queries
    test_queries = [
        "What's the weather in Seattle?",
        "Calculate 15 * 28",
        "Tell me about user 2",
    ]
    
    for query in test_queries:
        print(f"You: {query}")
        print("Assistant: ", end='', flush=True)
        
        try:
            response = agent.generate(query)
            print(response)
            print()
        except Exception as e:
            print(f"\nError: {e}\n")
    
    # Interactive mode
    print("=" * 50)
    print("Interactive mode - Type your own queries (or 'quit' to exit)")
    print("=" * 50)
    print()
    
    try:
        while True:
            user_input = input("You: ")
            
            if user_input.lower() in ["quit", "exit", ""]:
                break
            
            print("Assistant: ", end='', flush=True)
            
            try:
                response = agent.generate(user_input)
                print(response)
                print()
            except Exception as e:
                print(f"\nError: {e}\n")
    
    except KeyboardInterrupt:
        print("\n\nInterrupted by user")
    
    finally:
        print("\nExample complete!")
        agent.close()
        model.close()
    
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: examples/tool_calling.cpp
/**
 * @file tool_calling.cpp
 * @brief Tool calling example using luup-agent
 * 
 * This example demonstrates:
 * - Registering custom tools
 * - Tool callback implementation
 * - Automatic tool invocation by the agent
 * 
 * Note: This is a stub that will work once Phase 2 is complete
 */

#include <luup_agent.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

// Example tool: Get current weather
char* get_weather_callback(const char* params_json, void* user_data) {
    printf("  [Tool called: get_weather with params: %s]\n", params_json);
    
    // In a real implementation, you would:
    // 1. Parse params_json to extract the city
    // 2. Call a weather API
    // 3. Return formatted JSON result
    
    // For this example, return mock data
    const char* result = "{\"temperature\": 72, \"condition\": \"sunny\", \"humidity\": 45}";
    return strdup(result);
}

// Example tool: Calculate
char* calculate_callback(const char* params_json, void* user_data) {
    printf("  [Tool called: calculate with params: %s]\n", params_json);
    
    // Mock calculation result
    const char* result = "{\"result\": 42}";
    return strdup(result);
}

int main(int argc, char** argv) {
    printf("luup-agent Tool Calling Example\n");
    printf("Version: %s\n\n", luup_version());
    
    if (argc < 2) {
        printf("Usage: %s <path-to-model.gguf>\n", argv[0]);
        return 1;
    }
    
    const char* model_path = argv[1];
    
    // Create model
    luup_model_config model_config = {
        .path = model_path,
        .gpu_layers = -1,
        .context_size = 2048,
        .threads = 0,
        .api_key = nullptr,
        .api_base_url = nullptr
    };
    
    printf("Creating model...\n");
    luup_model* model = luup_model_create_local(&model_config);
    if (!model) {
        fprintf(stderr, "Error: %s\n", luup_get_last_error());
        return 1;
    }
    
    // Create agent with tool calling enabled
    luup_agent_config agent_config = {
        .model = model,
        .system_prompt = "You are a helpful assistant with access to tools. "
                        "Use the get_weather tool to check weather and "
                        "the calculate tool for math operations.",
        .temperature = 0.7f,
        .max_tokens = 512,
        .enable_tool_calling = true,  // Enable tool calling
        .enable_history_management = true
    };
    
    printf("Creating agent...\n");
    luup_agent* agent = luup_agent_create(&agent_config);
    if (!agent) {
        fprintf(stderr, "Error: %s\n", luup_get_last_error());
        luup_model_destroy(model);
        return 1;
    }
    
    // Register weather tool
    luup_tool weather_tool = {
        .name = "get_weather",
        .description = "Get current weather for a city",
        .parameters_json = 
            "{"
            "  \"type\": \"object\","
            "  \"properties\": {"
            "    \"city\": {"
            "      \"type\": \"string\","
            "      \"description\": \"The city name\""
            "    },"
            "    \"units\": {"
            "      \"type\": \"string\","
            "      \"enum\": [\"celsius\", \"fahrenheit\"],"
            "      \"description\": \"Temperature units\""
            "    }"
            "  },"
            "  \"required\": [\"city\"]"
            "}"
    };
    
    printf("Registering weather tool...\n");
    if (luup_agent_register_tool(agent, &weather_tool, get_weather_callback, nullptr) != LUUP_SUCCESS) {
        fprintf(stderr, "Error registering tool: %s\n", luup_get_last_error());
    }
    
    // Register calculator tool
    luup_tool calc_tool = {
        .name = "calculate",
        .description = "Perform mathematical calculations",
        .parameters_json =
            "{"
            "  \"type\": \"object\","
            "  \"properties\": {"
            "    \"expression\": {"
            "      \"type\": \"string\","
            "      \"description\": \"Mathematical expression to evaluate\""
            "    }"
            "  },"
            "  \"required\": [\"expression\"]"
            "}"
    };
    
    printf("Registering calculator tool...\n");
    if (luup_agent_register_tool(agent, &calc_tool, calculate_callback, nullptr) != LUUP_SUCCESS) {
        fprintf(stderr, "Error registering tool: %s\n", luup_get_last_error());
    }
    
    printf("\nTools registered successfully!\n\n");
    printf("==========================================\n");
    printf("Try these example queries:\n");
    printf("- What's the weather in Seattle?\n");
    printf("- Calculate 15 * 28\n");
    printf("- What's the weather in Tokyo and is it warmer than London?\n");
    printf("==========================================\n\n");
    
    // Test queries
    const char* test_queries[] = {
        "What's the weather in Seattle?",
        "Calculate 42 * 13",
        nullptr
    };
    
    for (int i = 0; test_queries[i] != nullptr; i++) {
        printf("You: %s\n", test_queries[i]);
        printf("Assistant: ");
        fflush(stdout);
        
        char* response = luup_agent_generate(agent, test_queries[i]);
        if (response) {
            printf("%s\n\n", response);
            luup_free_string(response);
        } else {
            fprintf(stderr, "Error: %s\n\n", luup_get_last_error());
        }
    }
    
    // Cleanup
    luup_agent_destroy(agent);
    luup_model_destroy(model);
    
    printf("Example complete!\n");
    return 0;
}

